src/mito2/src/read/range.rs

```diff
@@ -334,6 +341,160 @@ fn maybe_split_ranges_for_seq_scan(ranges: Vec<RangeMeta>) -> Vec<RangeMeta> {
+                row_group_index,
+                row_selection.clone(),
+            ));
+        } else {
+            // Scans all row groups.
+            ranges.extend(
+                self.row_groups
+                    .iter()
+                    .map(|(row_group_index, row_selection)| {
+                        FileRange::new(context.clone(), *row_group_index, row_selection.clone())
+                    }),
+            );
+        }
+    }
+}
+
+/// Builder to create mem ranges.
+pub(crate) struct MemRangeBuilder {
+    /// Ranges of a memtable.
+    row_groups: BTreeMap<usize, MemtableRange>,
+}
+
+impl MemRangeBuilder {
+    /// Builds a mem range builder from row groups.
+    pub(crate) fn new(row_groups: BTreeMap<usize, MemtableRange>) -> Self {
+        Self { row_groups }
+    }
+
+    /// Builds mem ranges to read in the memtable.
+    /// Negative `row_group_index` indicates all row groups.
+    fn build_ranges(&self, row_group_index: i64, ranges: &mut SmallVec<[MemtableRange; 2]>) {
+        if row_group_index >= 0 {
+            let row_group_index = row_group_index as usize;
+            // Scans one row group.
+            let Some(range) = self.row_groups.get(&row_group_index) else {
+                return;
+            };
+            ranges.push(range.clone());
+        } else {
+            ranges.extend(self.row_groups.values().cloned());
+        }
+    }
+}
+
+/// List to manages the builders to create file ranges.
+/// Each scan partition should have its own list. Mutex inside this list is used to allow moving
+/// the list to different streams in the same partition.
+pub(crate) struct RangeBuilderList {
+    num_memtables: usize,
+    mem_builders: Mutex<Vec<Option<MemRangeBuilder>>>,
+    file_builders: Mutex<Vec<Option<Arc<FileRangeBuilder>>>>,
+}
+
+impl RangeBuilderList {
+    /// Creates a new [ReaderBuilderList] with the given number of memtables and files.
+    pub(crate) fn new(num_memtables: usize, num_files: usize) -> Self {
+        let mem_builders = (0..num_memtables).map(|_| None).collect();
+        let file_builders = (0..num_files).map(|_| None).collect();
+        Self {
+            num_memtables,
+            mem_builders: Mutex::new(mem_builders),
+            file_builders: Mutex::new(file_builders),
+        }
+    }
+
+    /// Builds file ranges to read the row group at `index`.
+    pub(crate) async fn build_file_ranges(
+        &self,
+        input: &ScanInput,
+        index: RowGroupIndex,
+        reader_metrics: &mut ReaderMetrics,
+    ) -> Result<SmallVec<[FileRange; 2]>> {
+        let mut ranges = SmallVec::new();
+        let file_index = index.index - self.num_memtables;
+        let builder_opt = self.get_file_builder(file_index);
+        match builder_opt {
+            Some(builder) => builder.build_ranges(index.row_group_index, &mut ranges),
+            None => {
+                let builder = input.prune_file(file_index, reader_metrics).await?;
```

Comment:
Is the key point of this PR is to remove the std::mutex across the await point here so that top node in merge reader will not be affected by some other node silently holding the lock without being polled?

Reply:
Yes. The removed mutex is an async tokio mutex. It's better to avoid a partition waiting for the execution of another partition. It's fine to use mutex inside a partition if it doesn't need to wait for other partitions.